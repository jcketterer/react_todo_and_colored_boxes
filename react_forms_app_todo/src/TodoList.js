import React, { useState } from 'react'
import Todo from './Todo'
import TodoListForm from './TodoListForm'

const TodoList = () => {
  const [todos, setTodos] = useState([])

  const createTodo = newTodo => {
    setTodos(todos => [...todos, newTodo])
  }

  const updateTodo = (id, updatedTodo) => {
    setTodos(todos =>
      todos.map(todo => (todo.id === id ? { ...todo, task: updatedTodo } : todo))
    )
  }

  const remove = id => {
    setTodos(todos => todos.filter(todo => todo.id !== id))
  }

  const todoAttributes = todos.map(todo => (
    <Todo remove={remove} key={todo.id} id={todo.id} task={todo.task} update={updateTodo} />
  ))

  return (
    <div>
      <TodoListForm createTodo={createTodo} />
      <ul>{todoAttributes}</ul>
    </div>
  )
}

export default TodoList
