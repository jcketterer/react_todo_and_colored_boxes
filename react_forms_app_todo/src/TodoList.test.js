import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import TodoList from './TodoList'

const addTodo = (task = 'write all the tests') => {
  const input = screen.getByLabelText('Task:')
  fireEvent.change(input, { target: { value: task } })
  const submitBtn = screen.getByText('Add Todo!')
  fireEvent.click(submitBtn)
}

it('should render without breaking', () => {
  render(<TodoList />)
})

it('matches snap', () => {
  const { asFragment } = render(<TodoList />)
  expect(asFragment()).toMatchSnapshot()
})

it('should add a todo', () => {
  render(<TodoList />)
  addTodo()
  expect(screen.getByLabelText('Task:')).toHaveValue('')
  expect(screen.getByText('write all the tests')).toBeInTheDocument()
  expect(screen.getByText('Edit')).toBeInTheDocument()
  expect(screen.getByText('Remove')).toBeInTheDocument()
})

it('should edit a todo', () => {
  render(<TodoList />)
  addTodo()

  fireEvent.click(screen.getByText('Edit'))
  const input = screen.getByDisplayValue('write all the tests')
  fireEvent.change(input, { target: { value: 'sleep!' } })
  fireEvent.click(screen.getByText('Update!'))

  expect(screen.getByText('sleep!')).toBeInTheDocument()
  expect(screen.queryByText('write all the tests')).not.toBeInTheDocument()
})

it('should delete a todo', () => {
  render(<TodoList />)
  addTodo()

  fireEvent.click(screen.getByText('Remove'))
  expect(screen.queryByText('write all the tests')).not.toBeInTheDocument()
})
