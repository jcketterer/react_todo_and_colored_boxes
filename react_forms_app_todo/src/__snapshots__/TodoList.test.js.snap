// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snap 1`] = `
<DocumentFragment>
  <div>
    <div>
      <form>
        <label
          for="task"
        >
          Task: 
        </label>
        <input
          id="task"
          name="task"
          type="text"
          value=""
        />
        <button>
          Add Todo!
        </button>
      </form>
    </div>
    <ul />
  </div>
</DocumentFragment>
`;
